src/plugins/kibana_utils/public/state_management/url/kbn_url_tracker.ts

```diff
@@ -103,6 +104,12 @@ export function createKbnUrlTracker({
    * History object to use to track url changes. If this isn't provided, a local history instance will be created.
    */
   history?: History;
+
+  /**
+   * Lazily retrieve history instance
+   */
+  getHistory?: () => History;
```

Comment:
Because of this: https://github.com/elastic/kibana/pull/62761#discussion_r410224810

When `createHashHistory()` is created it adds `#` into the url. When discover will be launched as np app if to create it in setup it will change the url to `#` even though it is could be not needed (if we actually in the app which doesn't have hash in url). 

So went with this function to lazily create history when discover will actually need it for the first time, which will be when it is mounted. 

This should go away if/when discover is moved to use ScopedHistory. 

@sulemanof, please correct me if I am wrong with the reasoning.